### PR TITLE
Update compare_set implementation for FileStore and HashStore

### PR DESCRIPTION
**Pull request by H-Huang**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #57203 Update compare_set docs
* #57191 Add compare_set to trampoline class, add typing and formatting
* **#57175 Update compare_set implementation for FileStore and HashStore**

Update other Store implementations to add the value when current value is empty to match the amendment made to TCPStore (#55636). Added test to cover this case.

Test:
`pytest -vs test/distributed/test_c10d_common.py -k compare_set`

Differential Revision: [D28069380](https://our.internmc.facebook.com/intern/diff/D28069380)

### CODE REVIEW

File: torch/lib/c10d/FileStore.cpp [hunk 1]
```diff
@@ -140,7 +140,8 @@ class File {
     const auto start = std::chrono::steady_clock::now();
     while (true) {
 #ifdef _WIN32
-      fd_ = syscall(std::bind(::open, path.c_str(), flags | _O_BINARY, _S_IREAD | _S_IWRITE));
+      fd_ = syscall(std::bind(
+          ::open, path.c_str(), flags | _O_BINARY, _S_IREAD | _S_IWRITE));
```
Comment (H-Huang) [inline torch/lib/c10d/FileStore.cpp:144]: This change and above are purely formatting changes

File: torch/lib/c10d/FileStore.cpp [hunk 2]
```diff
@@ -310,9 +311,15 @@ std::vector<uint8_t> FileStore::compareSet(
   // it might be outdated
   pos_ = refresh(file, pos_, cache_);
   if (cache_.count(regKey) == 0) {
+    // if the key does not exist and current value is empty then set it
+    if (currentValue.empty()) {
```
Comment (cbalioglu) [inline torch/lib/c10d/FileStore.cpp:315]: Nit: Right now you have two places in the function where you write a new key/value pair to the file; one here at line 315 and the other at line 323.You can most likely rearrange the `if` conditions and consolidate them into one single place.

File: torch/lib/c10d/FileStore.cpp [hunk 3]
```diff
@@ -310,9 +311,15 @@ std::vector<uint8_t> FileStore::compareSet(
   // it might be outdated
   pos_ = refresh(file, pos_, cache_);
   if (cache_.count(regKey) == 0) {
+    // if the key does not exist and current value is empty then set it
```
Comment (cbalioglu) [inline torch/lib/c10d/FileStore.cpp:314]: Nit: Maybe you should re-word this comment. If the key does not exist, then there can't be any "current value". What you mean here with "current value" is the value of the `currentValue` parameter.

File: torch/lib/c10d/HashStore.cpp [hunk 4]
```diff
@@ -24,10 +24,14 @@ std::vector<uint8_t> HashStore::compareSet(
     const std::vector<uint8_t>& newValue) {
   std::unique_lock<std::mutex> lock(m_);
   auto it = map_.find(key);
-  if (it == map_.end()){
+  if (it == map_.end()) {
+    if (currentValue.empty()) {
+      map_[key] = newValue;
```
Comment (cbalioglu) [inline torch/lib/c10d/HashStore.cpp:29]: Nit: Same comment as above for `FileStore`.

**Review by cbalioglu — COMMENTED** [review]
Looks good to me. Just left a few minor comments. Let me know whether you want to address them; otherwise, I can approve it now.

File: torch/lib/c10d/FileStore.cpp (hunk 2 above)
Comment (H-Huang) [reply torch/lib/c10d/FileStore.cpp]: Thanks! Updated

File: torch/lib/c10d/FileStore.cpp (hunk 3 above)
Comment (H-Huang) [reply torch/lib/c10d/FileStore.cpp]: Updated the comment

**Review by cbalioglu — APPROVED** [review]
LGTM

**Pull request closed without merging**